python/ray/train/lightning/lightning_trainer.py

```diff
@@ -0,0 +1,298 @@
+    val_dataloaders = _prepare_dataloaders(
+        trainer_fit_config.get("val_dataloaders", None))
+
+    train_ray_dataset = session.get_dataset_shard("train")
+    val_ray_dataset = session.get_dataset_shard("val")
+    dataset_iter_config = config.pop(RAY_DATASET_ITER_CONFIG_KEY)
+    if train_ray_dataset:
+        if datamodule or train_dataloaders:
+            logger.warning(
+                "Using Ray Dataset as primary datasource, datamodule and dataloaders "
+                "specified in lightning_trainer_fit_params will be ignored."
+            )
+        datamodule = RayDataModule(
+            train_dataset=train_ray_dataset,
+            val_dataset=val_ray_dataset,
+            dataset_iter_config=dataset_iter_config
+        )
+
+    # Prepare Lightning Module
+    LightningModuleCls = config.pop(LIGHTNING_MODULE_KEY)
+    lightning_module_init_config = config.pop(LIGHTNING_MODULE_CONFIG_KEY)
+    lightning_module = LightningModuleCls(**lightning_module_init_config)
+
+    # Prepare Lightning Trainer
+    trainer_config = config.pop(LIGHTNING_TRAINER_CONFIG_KEY)
+    trainer_config["enable_progress_bar"] = False 
+    
+    
+    # Setup trainer's parallel devices
+    if trainer_config.get("accelerator", None) == "gpu":
+        current_device = ray.train.torch.get_device()
+        trainer_config["devices"] = [current_device.index]
+
+    # Setup ray cluster environment info
+    trainer_config["plugins"] = [plugin for plugin in trainer_config.get(
+        "plugins", []) if not isinstance(plugin, ClusterEnvironment)]
+    trainer_config["plugins"].append(RayEnvironment())
+
+    # Setup ddp strategy for ray orchestration
+    ddp_strategy_config = config.pop(DDP_STRATEGY_CONFIG_KEY)
+    trainer_config["strategy"] = RayDDPStrategy(**ddp_strategy_config)
+
+    # Prepare RayModelCheckpoint Callback
+    callbacks = trainer_config.get("callbacks", [])
+    trainer_config["callbacks"] = list(
+        filter(lambda callback: not isinstance(callback, ModelCheckpoint), callbacks))
+    if len(trainer_config["callbacks"]) < len(callbacks):
+        logger.warning(
+            "LightningTrainer only initialized one ModelCheckpoint callback based on"
+            " `model_checkpoint_config`. All other checkpoint callbacks are ignored."
+        )
+
+    # We always need a checkpoint callback to report metrics
+    trainer_config["enable_checkpointing"] = True 
+    # TODO(yunxuanx): Align with AIR Checkpoint Config
+    model_checkpoint_config = config.get(MODEL_CHECKPOINT_CONFIG_KEY)
+    ray_model_checkpoint = RayModelCheckpoint(**model_checkpoint_config)
+    trainer_config["callbacks"].append(ray_model_checkpoint)
+
+    trainer = pl.Trainer(**trainer_config)
+
+    # Restore the training from a previously interrupted/failed run.
+    checkpoint_path = None
+    checkpoint = session.get_checkpoint()
+    if checkpoint:
+        with checkpoint.as_directory() as ckpt_dir:
```

Comment:
`checkpoint.as_directory()` is just an ephemeral directory and should be treated as read-only.